python/ray/dag/tests/experimental/test_torch_tensor_dag.py

```diff
@@ -415,6 +420,212 @@ def test_torch_tensor_nccl_overlap_timed(ray_start_regular, overlap_gpu_communic
+
+    compiled_dag = dag.experimental_compile(
+        _overlap_gpu_communication=overlap_gpu_communication
+    )
+
+    elapses = []
+    start = time.monotonic()
+    for i in range(5):
+        start = time.monotonic()
+        ref = compiled_dag.execute(i)
+        result = ray.get(ref)
+        duration = time.monotonic() - start
+        elapses.append(duration)
+        assert (
+            result
+            == [(i * num_workers, coll_shape, dtype)] * num_workers
+            + [(i + comp_repeat, comp_shape, dtype)] * num_workers
+        )
+    duration = time.monotonic() - start
+    print(f"{overlap_gpu_communication=}, {duration=}")
+    for i, elapse in enumerate(elapses):
+        print(f"iteration {i=}, {elapse=}")
+
+    compiled_dag.teardown()
+
+
+@pytest.mark.skipif(not USE_GPU, reason="Skipping GPU Test")
+@pytest.mark.parametrize("overlap_gpu_communication", [False, True])
+def test_torch_tensor_nccl_overlap_p2p_and_collective(
+    ray_start_regular, overlap_gpu_communication
+):
+    assert (
+        sum(node["Resources"].get("GPU", 0) for node in ray.nodes()) >= 2
+    ), "This test requires at least 2 GPUs"
+
+    actor_cls = TorchTensorWorker.options(num_cpus=0, num_gpus=1)
+    num_workers = 2
+    workers = [actor_cls.remote() for _ in range(num_workers)]
+
+    dtype = torch.float16
+    coll_shape = (100_000_000,)
+    send_shape = (100_000_000,)
+    with InputNode() as inp:
+        coll_inputs = [worker.send.bind(coll_shape, dtype, inp) for worker in workers]
+        sends = [
+            worker.send.bind(send_shape, dtype, inp).with_tensor_transport("nccl")
+            for worker in workers
+        ]
+        recvs = [workers[0].recv.bind(sends[1]), workers[1].recv.bind(sends[0])]
+        coll_values = collective.allreduce.bind(coll_inputs)
+        coll_outputs = [
+            worker.recv.bind(coll) for worker, coll in zip(workers, coll_values)
+        ]
+        dag = MultiOutputNode(coll_outputs + recvs)
+
+    compiled_dag = dag.experimental_compile(
+        _overlap_gpu_communication=overlap_gpu_communication
+    )
+
+    elapses = []
+    start = time.monotonic()
+    for i in range(5):
+        start = time.monotonic()
+        ref = compiled_dag.execute(i)
+        result = ray.get(ref)
+        duration = time.monotonic() - start
+        elapses.append(duration)
+        assert (
+            result
+            == [(i * num_workers, coll_shape, dtype)] * num_workers
+            + [(i, send_shape, dtype)] * num_workers
+        )
+    duration = time.monotonic() - start
+    print(f"{overlap_gpu_communication=}, {duration=}")
+    for i, elapse in enumerate(elapses):
+        print(f"iteration {i=}, {elapse=}")
+
+    compiled_dag.teardown()
+
+
+@pytest.mark.skipif(not USE_GPU, reason="Skipping GPU Test")
+@pytest.mark.parametrize("overlap_gpu_communication", [False, True])
+def test_torch_tensor_nccl_overlap_send_future_across_actors(
```

Comment:
This test seems a bit complicated / unrelated compared to the stated goal? Is there a simpler test that can be run? Or a unit test?